File: frontend/src/types/raymondsIndex.ts
/**
 * RaymondsIndex 타입 정의
 *
 * 자본 배분 효율성 지수 관련 타입
 */

// ============================================================================
// 기본 타입
// ============================================================================

/** RaymondsIndex 등급 */
export type RaymondsIndexGrade = 'A+' | 'A' | 'B' | 'C' | 'D'

/** Sub-Index 점수 */
export interface SubIndexScores {
  /** Capital Efficiency Index (자본효율성 지수) - 20% */
  cei: number
  /** Reinvestment Intensity Index (재투자 강도 지수) - 35% ⭐ 핵심 */
  rii: number
  /** Cash Governance Index (현금 거버넌스 지수) - 25% */
  cgi: number
  /** Momentum Alignment Index (모멘텀 정렬 지수) - 20% */
  mai: number
}

/** 핵심 지표 */
export interface CoreMetrics {
  /** 투자괴리율 (%) = Cash CAGR - CAPEX Growth */
  investmentGap: number
  /** 현금 증가율 CAGR (%) */
  cashCagr: number
  /** CAPEX 증가율 (%) */
  capexGrowth: number
  /** 유휴현금비율 (%) */
  idleCashRatio: number
  /** 자산회전율 (회) */
  assetTurnover: number
  /** 재투자율 (%) */
  reinvestmentRate: number
  /** 주주환원율 (%) */
  shareholderReturn: number
}

/** 위험 신호 */
export interface RiskFlag {
  code: string
  message: string
  severity: 'red' | 'yellow'
}

// ============================================================================
// API 응답 타입
// ============================================================================

/** API 응답: RaymondsIndex */
export interface ApiRaymondsIndex {
  id: string
  company_id: string
  company_name?: string
  company_ticker?: string
  company_market?: string
  fiscal_year: number
  calculation_date: string | null
  total_score: number
  grade: RaymondsIndexGrade

  // Sub-Index
  cei_score: number | null
  rii_score: number | null
  cgi_score: number | null
  mai_score: number | null

  // Core Metrics
  investment_gap: number | null
  cash_cagr: number | null
  capex_growth: number | null
  idle_cash_ratio: number | null
  asset_turnover: number | null
  reinvestment_rate: number | null
  shareholder_return: number | null

  // Flags
  red_flags: string[]
  yellow_flags: string[]

  // Interpretation
  verdict: string | null
  key_risk: string | null
  recommendation: string | null
  watch_trigger: string | null

  // Metadata
  data_quality_score: number | null
  rank?: number
}

/** API 응답: 랭킹 목록 */
export interface ApiRaymondsIndexRanking {
  total: number
  offset: number
  limit: number
  rankings: ApiRaymondsIndex[]
}

/** API 응답: 히스토리 */
export interface ApiRaymondsIndexHistory {
  company_id: string
  company_name: string
  history: ApiRaymondsIndex[]
}

/** API 응답: 통계 */
export interface ApiRaymondsIndexStatistics {
  total_companies: number
  average_score: number
  min_score: number
  max_score: number
  average_investment_gap: number
  grade_distribution: Record<RaymondsIndexGrade, number>
  year: number | null
}

// ============================================================================
// 프론트엔드 타입
// ============================================================================

/** 프론트엔드용 RaymondsIndex */
export interface RaymondsIndexData {
  id: string
  companyId: string
  companyName?: string
  companyTicker?: string
  companyMarket?: string
  fiscalYear: number
  calculationDate: string | null

  // 종합
  totalScore: number
  grade: RaymondsIndexGrade

  // Sub-Index
  subIndexScores: SubIndexScores

  // Core Metrics
  coreMetrics: CoreMetrics

  // Flags
  redFlags: string[]
  yellowFlags: string[]

  // Interpretation
  verdict: string | null
  keyRisk: string | null
  recommendation: string | null
  watchTrigger: string | null

  // Metadata
  dataQualityScore: number | null
  rank?: number
}

// ============================================================================
// 유틸리티 함수
// ============================================================================

/** API 응답을 프론트엔드 타입으로 변환 */
export function mapApiToRaymondsIndex(api: ApiRaymondsIndex): RaymondsIndexData {
  return {
    id: api.id,
    companyId: api.company_id,
    companyName: api.company_name,
    companyTicker: api.company_ticker,
    companyMarket: api.company_market,
    fiscalYear: api.fiscal_year,
    calculationDate: api.calculation_date,
    totalScore: api.total_score,
    grade: api.grade,
    subIndexScores: {
      cei: api.cei_score ?? 0,
      rii: api.rii_score ?? 0,
      cgi: api.cgi_score ?? 0,
      mai: api.mai_score ?? 0,
    },
    coreMetrics: {
      investmentGap: api.investment_gap ?? 0,
      cashCagr: api.cash_cagr ?? 0,
      capexGrowth: api.capex_growth ?? 0,
      idleCashRatio: api.idle_cash_ratio ?? 0,
      assetTurnover: api.asset_turnover ?? 0,
      reinvestmentRate: api.reinvestment_rate ?? 0,
      shareholderReturn: api.shareholder_return ?? 0,
    },
    redFlags: api.red_flags || [],
    yellowFlags: api.yellow_flags || [],
    verdict: api.verdict,
    keyRisk: api.key_risk,
    recommendation: api.recommendation,
    watchTrigger: api.watch_trigger,
    dataQualityScore: api.data_quality_score,
    rank: api.rank,
  }
}

/** 등급별 색상 */
export function getGradeColor(grade: RaymondsIndexGrade): string {
  const colors: Record<RaymondsIndexGrade, string> = {
    'A+': '#22c55e', // green-500
    'A': '#84cc16',  // lime-500
    'B': '#eab308',  // yellow-500
    'C': '#f97316',  // orange-500
    'D': '#ef4444',  // red-500
  }
  return colors[grade] || '#6b7280'
}

/** 등급별 배경색 */
export function getGradeBgColor(grade: RaymondsIndexGrade): string {
  const colors: Record<RaymondsIndexGrade, string> = {
    'A+': 'bg-green-100 text-green-800',
    'A': 'bg-lime-100 text-lime-800',
    'B': 'bg-yellow-100 text-yellow-800',
    'C': 'bg-orange-100 text-orange-800',
    'D': 'bg-red-100 text-red-800',
  }
  return colors[grade] || 'bg-gray-100 text-gray-800'
}

// ============================================================================
// 랭킹 페이지용 타입
// ============================================================================

/** 랭킹 페이지 아이템 */
export interface RaymondsIndexRankingItem {
  company_id: string
  company_name: string
  stock_code?: string
  market?: string
  total_score: number
  grade: RaymondsIndexGrade
  investment_gap: number
  cei_score?: number
  rii_score?: number
  cgi_score?: number
  mai_score?: number
  fiscal_year: number
}

/** 통계 타입 (랭킹 페이지용) */
export interface RaymondsIndexStatistics {
  total_companies: number
  average_score: number
  min_score?: number
  max_score?: number
  average_investment_gap: number
  grade_distribution: Record<string, number>
  year?: number
}

/** 투자괴리율 해석 */
export function getInvestmentGapInterpretation(gap: number): {
  status: 'good' | 'warning' | 'danger'
  message: string
} {
  if (gap >= -5 && gap <= 5) {
    return { status: 'good', message: '균형 잡힌 자본 배분' }
  } else if (gap > 5 && gap <= 15) {
    return { status: 'warning', message: '현금 축적 경향' }
  } else if (gap > 15) {
    return { status: 'danger', message: '과다 현금 축적 - 투자 부족' }
  } else if (gap < -5 && gap >= -15) {
    return { status: 'warning', message: '적극적 투자 중' }
  } else {
    return { status: 'danger', message: '과잉 투자 - 현금 고갈 위험' }
  }
}

File: backend/app/models/raymonds_index.py
"""
RaymondsIndex 모델 - 자본 배분 효율성 지수
"""
from sqlalchemy import Column, String, Integer, Date, DateTime, ForeignKey, UniqueConstraint, Index, Numeric, Text
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import uuid

from app.database import Base


class RaymondsIndex(Base):
    """
    RaymondsIndex v2.1 계산 결과 테이블

    자본 배분 효율성을 측정하는 종합 지수로,
    4개의 Sub-Index로 구성됨:
    - CEI: Capital Efficiency Index (20%) - 자본 효율성
      → 자산회전율(25%) + 유형자산효율성(20%) + 현금수익률(20%) + ROIC(25%) + 추세(10%)
    - RII: Reinvestment Intensity Index (35%) ⭐ 핵심 - 재투자 강도
      → CAPEX강도(30%) + 투자괴리율(30%) + 재투자율(25%) + 지속성(15%)
    - CGI: Cash Governance Index (25%) ⭐ 핵심 - 현금 거버넌스
      → 현금활용도(20%) + 자금조달효율성(25%) + 주주환원균형(20%) + 현금적정성(15%) + 부채건전성(20%)
    - MAI: Momentum Alignment Index (20%) - 모멘텀 정합성
      → 매출-투자동조성(30%) + 이익품질(25%) + 투자지속성(20%) + 성장투자비율(15%) + FCF추세(10%)

    v2.1 주요 변경 (v2.0 대비):
    - 투자괴리율: 현금 CAGR - CAPEX 성장률 (v2.0: 재투자율 기반)
    - 재투자율: CAPEX / OCF (v2.0: CAPEX / 영업이익)
    - 등급 기준 완화 (A+: 90→88, A: 85→80, etc.)
    - 신규 지표: 유형자산효율성, 현금수익률, 부채건전성, 성장투자비율

    특별 규칙:
    - 현금-유형자산 비율 > 30:1 → 최대 B-
    - 조달자금 전환율 < 30% → 최대 B-
    - 단기금융상품비율 > 65% + CAPEX 감소 → 최대 B
    - 위 조건 2개 이상 해당 → 최대 C+
    """
    __tablename__ = "raymonds_index"

    # Primary Key
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    # Foreign Key
    company_id = Column(UUID(as_uuid=True), ForeignKey('companies.id', ondelete='CASCADE'), nullable=False)

    # Calculation metadata
    calculation_date = Column(Date, nullable=False, server_default=func.current_date())
    fiscal_year = Column(Integer, nullable=False)

    # ═══════════════════════════════════════════════════════════════
    # 종합 점수
    # ═══════════════════════════════════════════════════════════════
    total_score = Column(Numeric(5, 2), nullable=False)  # 0-100
    grade = Column(String(5), nullable=False)            # A+, A, B, C, D

    # ═══════════════════════════════════════════════════════════════
    # Sub-Index 점수 (각 0-100)
    # ═══════════════════════════════════════════════════════════════
    cei_score = Column(Numeric(5, 2), nullable=True)     # Capital Efficiency Index (20%)
    rii_score = Column(Numeric(5, 2), nullable=True)     # Reinvestment Intensity Index (35%) ⭐
    cgi_score = Column(Numeric(5, 2), nullable=True)     # Cash Governance Index (25%)
    mai_score = Column(Numeric(5, 2), nullable=True)     # Momentum Alignment Index (20%)

    # ═══════════════════════════════════════════════════════════════
    # 핵심 지표 (기존)
    # ═══════════════════════════════════════════════════════════════
    investment_gap = Column(Numeric(6, 2), nullable=True)     # 투자괴리율 (%) = Cash CAGR - CAPEX Growth
    cash_cagr = Column(Numeric(6, 2), nullable=True)          # 현금 증가율 CAGR (%)
    capex_growth = Column(Numeric(6, 2), nullable=True)       # CAPEX 증가율 (%)
    idle_cash_ratio = Column(Numeric(5, 2), nullable=True)    # 유휴현금비율 (%)
    asset_turnover = Column(Numeric(5, 3), nullable=True)     # 자산회전율 (회)
    reinvestment_rate = Column(Numeric(5, 2), nullable=True)  # 재투자율 (%)
    shareholder_return = Column(Numeric(5, 2), nullable=True) # 주주환원율 (%)

    # ═══════════════════════════════════════════════════════════════
    # v4.0 기존 지표
    # ═══════════════════════════════════════════════════════════════
    cash_tangible_ratio = Column(Numeric(10, 2), nullable=True)    # 현금-유형자산 증가비율 (X:1)
    fundraising_utilization = Column(Numeric(5, 2), nullable=True) # 조달자금 투자전환율 (%)
    short_term_ratio = Column(Numeric(5, 2), nullable=True)        # 단기금융상품 비율 (%)
    capex_trend = Column(String(20), nullable=True)                # CAPEX 추세 (increasing/stable/decreasing)
    roic = Column(Numeric(6, 2), nullable=True)                    # 투하자본수익률 ROIC (%)
    capex_cv = Column(Numeric(5, 3), nullable=True)                # CAPEX 변동계수 (투자 지속성)
    violation_count = Column(Integer, default=0)                   # 특별규칙 위반 개수

    # ═══════════════════════════════════════════════════════════════
    # v2.0/v2.1 지표
    # ═══════════════════════════════════════════════════════════════
    investment_gap_v2 = Column(Numeric(6, 2), nullable=True)       # 투자괴리율 v2 (레거시: 초기2년 - 최근2년 재투자율)
    investment_gap_v21 = Column(Numeric(6, 2), nullable=True)      # 투자괴리율 v2.1 ⭐핵심 (현금 CAGR - CAPEX 성장률)
    rd_intensity = Column(Numeric(5, 2), nullable=True)            # R&D 강도 - 미사용 (데이터 없음)
    ebitda = Column(Numeric(20, 0), nullable=True)                 # EBITDA
    debt_to_ebitda = Column(Numeric(6, 2), nullable=True)          # 부채/EBITDA (CGI용)
    cash_utilization = Column(Numeric(5, 2), nullable=True)        # 현금 활용도 (%)
    industry_sector = Column(String(50), nullable=True)            # 업종 분류
    weight_adjustment = Column(JSONB, nullable=True)               # 업종별 가중치 조정 내역

    # ═══════════════════════════════════════════════════════════════
    # v2.1 신규 지표
    # ═══════════════════════════════════════════════════════════════
    tangible_efficiency = Column(Numeric(6, 3), nullable=True)     # 유형자산 효율성 (매출/유형자산)
    cash_yield = Column(Numeric(6, 2), nullable=True)              # 현금 수익률 (영업이익/총현금 %)
    growth_investment_ratio = Column(Numeric(5, 2), nullable=True) # 성장 투자 비율 (성장CAPEX/총CAPEX %)

    # ═══════════════════════════════════════════════════════════════
    # 위험 신호 (JSONB 배열)
    # ═══════════════════════════════════════════════════════════════
    red_flags = Column(JSONB, default=list)      # 위험 신호 배열
    yellow_flags = Column(JSONB, default=list)   # 주의 신호 배열

    # ═══════════════════════════════════════════════════════════════
    # 해석
    # ═══════════════════════════════════════════════════════════════
    verdict = Column(String(200), nullable=True)       # 한 줄 요약
    key_risk = Column(Text, nullable=True)             # 핵심 리스크 설명
    recommendation = Column(Text, nullable=True)       # 투자자 권고
    watch_trigger = Column(Text, nullable=True)        # 재검토 시점

    # ═══════════════════════════════════════════════════════════════
    # 메타데이터
    # ═══════════════════════════════════════════════════════════════
    data_quality_score = Column(Numeric(3, 2), nullable=True)  # 데이터 품질 점수 (0-1)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    # Relationships
    company = relationship("Company", back_populates="raymonds_indices")

    # Constraints
    __table_args__ = (
        UniqueConstraint('company_id', 'fiscal_year', name='uq_raymonds_index'),
        Index('idx_ri_company', 'company_id'),
        Index('idx_ri_year', 'fiscal_year'),
        Index('idx_ri_score', 'total_score'),
        Index('idx_ri_grade', 'grade'),
        Index('idx_ri_investment_gap', 'investment_gap'),
    )

    def __repr__(self):
        return f"<RaymondsIndex(company_id={self.company_id}, year={self.fiscal_year}, score={self.total_score}, grade={self.grade})>"

    def to_dict(self):
        """API 응답용 딕셔너리 변환"""
        return {
            "id": str(self.id),
            "company_id": str(self.company_id),
            "fiscal_year": self.fiscal_year,
            "calculation_date": self.calculation_date.isoformat() if self.calculation_date else None,
            "total_score": float(self.total_score) if self.total_score else None,
            "grade": self.grade,
            # Sub-Index 점수
            "cei_score": float(self.cei_score) if self.cei_score else None,
            "rii_score": float(self.rii_score) if self.rii_score else None,
            "cgi_score": float(self.cgi_score) if self.cgi_score else None,
            "mai_score": float(self.mai_score) if self.mai_score else None,
            # 기존 핵심 지표
            "investment_gap": float(self.investment_gap) if self.investment_gap else None,
            "cash_cagr": float(self.cash_cagr) if self.cash_cagr else None,
            "capex_growth": float(self.capex_growth) if self.capex_growth else None,
            "idle_cash_ratio": float(self.idle_cash_ratio) if self.idle_cash_ratio else None,
            "asset_turnover": float(self.asset_turnover) if self.asset_turnover else None,
            "reinvestment_rate": float(self.reinvestment_rate) if self.reinvestment_rate else None,
            "shareholder_return": float(self.shareholder_return) if self.shareholder_return else None,
            # v4.0 기존 지표
            "cash_tangible_ratio": float(self.cash_tangible_ratio) if self.cash_tangible_ratio else None,
            "fundraising_utilization": float(self.fundraising_utilization) if self.fundraising_utilization else None,
            "short_term_ratio": float(self.short_term_ratio) if self.short_term_ratio else None,
            "capex_trend": self.capex_trend,
            "roic": float(self.roic) if self.roic else None,
            "capex_cv": float(self.capex_cv) if self.capex_cv else None,
            "violation_count": self.violation_count or 0,
            # v2.0/v2.1 지표
            "investment_gap_v2": float(self.investment_gap_v2) if self.investment_gap_v2 else None,
            "investment_gap_v21": float(self.investment_gap_v21) if self.investment_gap_v21 else None,
            "cash_utilization": float(self.cash_utilization) if self.cash_utilization else None,
            "industry_sector": self.industry_sector,
            "weight_adjustment": self.weight_adjustment,
            # v2.1 신규 지표
            "tangible_efficiency": float(self.tangible_efficiency) if self.tangible_efficiency else None,
            "cash_yield": float(self.cash_yield) if self.cash_yield else None,
            "debt_to_ebitda": float(self.debt_to_ebitda) if self.debt_to_ebitda else None,
            "growth_investment_ratio": float(self.growth_investment_ratio) if self.growth_investment_ratio else None,
            # 위험 신호
            "red_flags": self.red_flags or [],
            "yellow_flags": self.yellow_flags or [],
            # 해석
            "verdict": self.verdict,
            "key_risk": self.key_risk,
            "recommendation": self.recommendation,
            "watch_trigger": self.watch_trigger,
            "data_quality_score": float(self.data_quality_score) if self.data_quality_score else None,
        }
